logs_analyzer/chapter1/scala/src/main/scala/com/databricks/apps/logs/chapter1/LogAnalyzer.scala

```diff
@@ -34,26 +34,15 @@ object LogAnalyzer extends App {
     contentSizes.max))
 
   // Compute Response Code to Count.
-  val responseCodeToCount: Array[(Int, Long)] = accessLogs
-    .map(_.responseCode -> 1L)
-    .reduceByKey(_ + _)
-    .take(100)
+  val responseCodeToCount: Array[(Int, Long)] = accessLogs.groupBy(_.responseCode).mapValues(_.size).take(100)
```

Comment:
`RDD.groupBy` creates an RDD of pairs: `RDD[Int, Iterable[ApacheAccessLog]]`. This is very inefficient in terms of memory consumption. Moreover, `groupBy` causes data reshuffling across executors, and in this case instances of `ApacheAccessLog` will be transferred via the network. This is very inefficient in terms of network consumption. Finally, each instance of `ApacheAccessLog` must be serialized and deserialized in order to transfer - additional CPU consumption.

On the other hand, `reduceByKey(_ + _)` stores only Long numbers for each key and transfers only Long numbers across executors. This is far more efficient.

Reply:
Thanks for such a detailed response @tashoyan. After looking at the RDD docs this performance issue is also mentioned. Will close this.